kaldb/src/main/java/com/slack/kaldb/clusterManager/RecoveryTaskAssignmentService.java

```diff
@@ -0,0 +1,227 @@
+      MeterRegistry meterRegistry) {
+    this.recoveryTaskMetadataStore = recoveryTaskMetadataStore;
+    this.recoveryNodeMetadataStore = recoveryNodeMetadataStore;
+    this.recoveryTaskAssignmentServiceConfig = recoveryTaskAssignmentServiceConfig;
+    this.meterRegistry = meterRegistry;
+
+    checkArgument(
+        recoveryTaskAssignmentServiceConfig.getEventAggregationSecs() > 0,
+        "eventAggregationSecs must be > 0");
+    // schedule configs checked as part of the AbstractScheduledService
+
+    recoveryTasksCreated = meterRegistry.counter(RECOVERY_TASKS_CREATED);
+    recoveryTasksFailed = meterRegistry.counter(RECOVERY_TASKS_FAILED);
+    recoveryAssignmentTimer = meterRegistry.timer(RECOVERY_TASK_ASSIGNMENT_TIMER);
+  }
+
+  @Override
+  protected synchronized void runOneIteration() {
+    if (pendingTask == null || pendingTask.getDelay(TimeUnit.SECONDS) <= 0) {
+      pendingTask =
+          executorService.schedule(
+              this::assignRecoveryTasksToNodes,
+              recoveryTaskAssignmentServiceConfig.getEventAggregationSecs(),
+              TimeUnit.SECONDS);
+    } else {
+      LOG.debug(
+          "Recovery task already queued for execution, will run in {} ms",
+          pendingTask.getDelay(TimeUnit.MILLISECONDS));
+    }
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting recovery task assignment service");
+    recoveryTaskMetadataStore.addListener(this::runOneIteration);
+    recoveryNodeMetadataStore.addListener(this::runOneIteration);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    executorService.shutdown();
+    LOG.info("Closed recovery task assignment service");
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        recoveryTaskAssignmentServiceConfig.getScheduleInitialDelayMins(),
+        recoveryTaskAssignmentServiceConfig.getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  /**
+   * Assigns the first recovery task needing assignment to the first available recovery node. No
+   * preference is given to specific recovery tasks, nor to executor nodes. This may result in
+   * over/under utilization of specific recovery nodes as well as indeterminate order of recovery
+   * task fulfillment.
+   *
+   * <p>If this method fails to successfully assign all the required tasks, the following iteration
+   * of this method would attempt to re-assign these until there are no more available tasks.
+   *
+   * @return The count of successfully assigned recovery tasks
+   */
+  protected int assignRecoveryTasksToNodes() {
+    Timer.Sample assignmentTimer = Timer.start(meterRegistry);
+
+    List<String> recoveryTasksAlreadyAssigned =
+        recoveryNodeMetadataStore
+            .getCached()
+            .stream()
+            .map((recoveryNodeMetadata -> recoveryNodeMetadata.recoveryTaskName))
+            .filter((recoveryTaskName) -> !recoveryTaskName.isEmpty())
```

Comment:
Instead of filtering by task name being empty would it be better if we filtered by RecoveryNodeState.FREE"?